### ISSUE TITLE
Documentation - Migration - dexie 2.x -> dexie 3.x

### ISSUE THREAD

**Issue by jokamax** [dfahlander/Dexie.js#807]
Hi,

Is there a way to have in documentation a practical example of migration from dexie 2.x to 3.x.
Mostly : starting (declaration/opening) and versioning ?

Thanks by advance,

Jonathan

**Comment by dfahlander** [comment]
Dexie 3.0 is highly backward compatible with 2.0 except for a rarely used property Transaction.tables, see https://dexie.org/docs/Deprecations. Declaration, opening and versioning is the same still, only that it also allows for more cases, such as renaming tables (See #713).

If you use Typescript, the typings is completely rewritten and we're trying to move away from namespaces to modules as much as possible. Kept the namespace-based type Dexie.Table<> in 3.0 but not all the other namespace based types, so if your typescript compiler complains, have a look in node_modules/dexie/dist/dexie.d.ts. If the type was accessed through the Dexie namespace before, you will find it as its own named export in Dexie 3.0.

**Comment by jokamax** [comment]
Thanks for explanation !

Then, this part define in release note 3.0.5-alpha is not mandatory  ?
```
db.use({
  stack: "dbcore", // The only stack supported so far.
  name: "MyMiddleware", // Optional name of your middleware
  create (downlevelDatabase) {
    // Return your own implementation of DBCore:
    return {
      // Copy default implementation.
      ...downlevelDatabase, 
      // Override table method
      table (tableName) {
        // Call default table method
        const downlevelTable = downlevelDatabase.table(tableName);
        // Derive your own table from it:
        return {
          // Copy default table implementation:
          ...downlevelTable,
          // Override the mutate method:
          mutate: req => {
            // Copy the request object
            const myRequest = {...req};
            // Do things before mutate, then
            // call downlevel mutate:
            return downlevelTable.mutate(myRequest).then(res => {
              // Do things after mutate
              const myResponse = {...res};
              // Then return your response:
              return myResponse;
            });
          }
        }
      }
    };
  }
});
```

**Comment by dfahlander** [comment]
No, this is just a new API that is available in 3.0.

**Comment by jokamax** [comment]
Thanks for the quick answer.
No more for me : THANKS
